app/controllers/AiModelController.scala

```diff
@@ -0,0 +1,224 @@
+    }
+  }
+
+  def readAiInferenceInfo(aiInferenceId: String): Action[AnyContent] = sil.SecuredAction.async { implicit request =>
+    {
+      for {
+        _ <- userService.assertIsSuperUser(request.identity)
+        aiInferenceIdValidated <- ObjectId.fromString(aiInferenceId)
+        aiInference <- aiInferenceDAO.findOne(aiInferenceIdValidated) ?~> "aiInference.notFound" ~> NOT_FOUND
+        jsResult <- aiInferenceService.publicWrites(aiInference, request.identity)
+      } yield Ok(jsResult)
+    }
+  }
+
+  def listAiModels: Action[AnyContent] = sil.SecuredAction.async { implicit request =>
+    {
+      for {
+        _ <- userService.assertIsSuperUser(request.identity)
+        aiModels <- aiModelDAO.findAll
+        jsResults <- Fox.serialCombined(aiModels)(aiModelService.publicWrites)
+      } yield Ok(Json.toJson(jsResults))
+    }
+  }
+
+  def listAiInferences: Action[AnyContent] = sil.SecuredAction.async { implicit request =>
+    {
+      for {
+        _ <- userService.assertIsSuperUser(request.identity)
+        aiInferences <- aiInferenceDAO.findAll
+        jsResults <- Fox.serialCombined(aiInferences)(inference =>
+          aiInferenceService.publicWrites(inference, request.identity))
+      } yield Ok(Json.toJson(jsResults))
+    }
+  }
+
+  def runTraining: Action[RunTrainingParameters] = sil.SecuredAction.async(validateJson[RunTrainingParameters]) {
+    implicit request =>
+      for {
+        _ <- userService.assertIsSuperUser(request.identity)
+        trainingAnnotations = request.body.trainingAnnotations
+        _ <- bool2Fox(trainingAnnotations.nonEmpty || request.body.workflowYaml.isDefined) ?~> "aiModel.training.zeroAnnotations"
+        firstAnnotationId <- trainingAnnotations.headOption.map(_.annotationId).toFox
+        annotation <- annotationDAO.findOne(firstAnnotationId)
+        dataset <- datasetDAO.findOne(annotation._dataset)
+        dataStore <- dataStoreDAO.findOneByName(dataset._dataStore) ?~> "dataStore.notFound"
+        _ <- Fox
+          .serialCombined(request.body.trainingAnnotations.map(_.annotationId))(annotationDAO.findOne) ?~> "annotation.notFound"
+        modelId = ObjectId.generate
+        organization <- organizationDAO.findOne(request.identity._organization)
+        jobCommand = JobCommand.train_model
+        commandArgs = Json.obj(
+          "training_annotations" -> Json.toJson(trainingAnnotations),
+          "organization_name" -> organization.name,
+          "model_id" -> modelId,
+          "workflow_yaml" -> request.body.workflowYaml
+        )
+        newTrainingJob <- jobService
+          .submitJob(jobCommand, commandArgs, request.identity, dataStore.name) ?~> "job.couldNotRunTrainModel"
+        existingAiModelsCount <- aiModelDAO.countByNameAndOrganization(request.body.name,
+                                                                       request.identity._organization)
+        _ <- bool2Fox(existingAiModelsCount == 0) ?~> "model.nameInUse"
```

Comment:
To me this reads like the job is first submitted and then the it is checked whether there is a name duplication. Shouldn't the name check be before the submision?

Reply:
good catch, yes!